src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -371,16 +368,62 @@ private async ValueTask<EvaluationEvent> EvaluateFeature<TContext>(string featur
                     await sessionManager.SetAsync(evaluationEvent.FeatureDefinition.Name, evaluationEvent.Enabled).ConfigureAwait(false);
                 }
 
-                if (evaluationEvent.FeatureDefinition.Telemetry != null &&
-                    evaluationEvent.FeatureDefinition.Telemetry.Enabled)
+                // Only add an activity event if telemetry is enabled for the feature
+                if (telemetryEnabled)
                 {
-                    PublishTelemetry(evaluationEvent, cancellationToken);
+                    AddEvaluationActivityEvent(evaluationEvent);
                 }
             }
 
             return evaluationEvent;
         }
 
+        private void AddEvaluationActivityEvent(EvaluationEvent evaluationEvent)
+        {
+            // Break if there is no reason to emit an event
+            if (Activity.Current == null || !Activity.Current.IsAllDataRequested)
```

Comment:
No need to call this method at all if this is the case. Typically better to avoid inserting immediate exit control-flow in a method that way at the call-site a method like `AddXyz` can always be expected to add something.

Reply:
Interesting perspective on this. What do you do when the method is called multiple times, and the condition is nontrivial (like this one)?

By not including the condition in the method itself, you not only increase duplication (by replicating the check) but also arguably make SRP worse by splitting a single "thing" into 2 places.

Reply:
In my opinion this condition is trivial. If the condition wasn't trivial an extension method may be helpful to check state.

I don't worry about duplication, I worry more about clarity.